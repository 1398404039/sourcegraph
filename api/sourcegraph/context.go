package sourcegraph

import "context"

type contextKey int

const (
	accessTokenKey contextKey = iota
)

// WithAccessToken returns a copy of the parent context that uses token
// as the access token for future API clients constructed using this
// context (with NewClientFromContext). It replaces (shadows) any
// previously set token in the context.
func WithAccessToken(parent context.Context, token string) context.Context {
	return context.WithValue(parent, accessTokenKey, token)
}

// AccessTokenFromContext returns the access token (if any) previously
// set in the context by WithAccessToken.
func AccessTokenFromContext(ctx context.Context) string {
	token, ok := ctx.Value(accessTokenKey).(string)
	if !ok {
		return ""
	}
	return token
}
