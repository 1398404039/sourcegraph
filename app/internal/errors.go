package internal

import (
	"errors"
	"fmt"
	"net/http"
	"reflect"

	opentracing "github.com/opentracing/opentracing-go"
	"github.com/opentracing/opentracing-go/ext"

	"gopkg.in/inconshreveable/log15.v2"
	"sourcegraph.com/sourcegraph/sourcegraph/api/sourcegraph/legacyerr"
	"sourcegraph.com/sourcegraph/sourcegraph/app/internal/tmpl"
	"sourcegraph.com/sourcegraph/sourcegraph/pkg/handlerutil"
	"sourcegraph.com/sourcegraph/sourcegraph/pkg/randstring"
	"sourcegraph.com/sourcegraph/sourcegraph/pkg/traceutil"
)

// ErrorHandler is a func that renders a custom error page for the
// given error (which was returned by an app handler).
type ErrorHandler func(http.ResponseWriter, *http.Request, error) error

var errorHandlers = map[reflect.Type]ErrorHandler{}

// RegisterErrorHandlerForType registers an error handler that
// HandleError delegates to, for all errors (returned by app handlers)
// of the same type as errorVal.
//
// For example, to delegate to handleFooError for all errors returned
// by app handlers of type "*foo", call
// RegisterErrorHandlerForType(&foo{}, handleFooError).
//
// These error handlers are matched BEFORE those registered with
// RegisterErrorHandler are matched.
//
// It should only be called at init time. It panics if there is
// already a handler registered for the type.
func RegisterErrorHandlerForType(errorVal interface{}, handler ErrorHandler) {
	t := reflect.TypeOf(errorVal)
	if _, present := errorHandlers[t]; present {
		return
	}
	errorHandlers[t] = handler
}

// UnauthorizedErrorHandler is the error handler that is called when
// an app handler returns an "unauthenticated" error (i.e.,
// legacyerr.ErrCode(err) == legacyerr.Unauthenticated).
//
// Currently it is set at init time by package authutil; this is
// necessary to avoid an import cycle.
var UnauthorizedErrorHandler ErrorHandler

// HandleError handles err, delegating to an error handler func
// previously registered with RegisterErrorHandler* if present, and
// otherwise displaying the standard error page.
func HandleError(resp http.ResponseWriter, req *http.Request, status int, err error) {
	origErr := err
	errorID := randstring.NewLen(6)

	if errH, present := errorHandlers[reflect.TypeOf(err)]; present {
		err = errH(resp, req, err)
		if err == nil {
			return
		}
	} else if UnauthorizedErrorHandler != nil && (legacyerr.ErrCode(err) == legacyerr.Unauthenticated || status == http.StatusUnauthorized) {
		log15.Debug("redirecting to login", "from", req.URL, "error", err, "error_id", errorID)
		err = UnauthorizedErrorHandler(resp, req, err)
		if err == nil {
			return
		}
	} else {
		// err is the error that occurred during an error handler, but
		// this if-branch is when there is no error handler, so set it
		// to nil.
		err = nil
	}

	if err == nil {
		err = origErr
	} else {
		// An error occurred during execution of the error handler.
		err = fmt.Errorf("during execution of error handler: %s (original error: %s)", err, origErr)
	}

	if status < 200 || status >= 400 {
		var spanURL string
		if span := opentracing.SpanFromContext(req.Context()); span != nil {
			ext.Error.Set(span, true)
			span.SetTag("err", err)
			spanURL = traceutil.SpanURL(span)
		}
		log15.Error("App HTTP handler error response", "method", req.Method, "request_uri", req.URL.RequestURI(), "status_code", status, "error", err, "error_id", errorID, "trace", spanURL)
	}

	// Handle panic during execution of error template.
	defer func() {
		if e := recover(); e != nil {
			log15.Error("panic during execution of error template", "error", e, "error_id", errorID, "func_name", "HandleError", "tmpl_name", "error/error.html")
			err := fmt.Errorf("panic during execution of error template (error id %v): %v", errorID, e)
			if !handlerutil.DebugMode {
				err = errPublicFacingErrorMessage
			}
			http.Error(resp, err.Error(), http.StatusInternalServerError)
		}
	}()

	// Display internal grpc error descriptions with full text (so it's not escaped).
	if legacyerr.ErrCode(err) == legacyerr.Internal {
		err = fmt.Errorf("internal error:\n\n%s", legacyerr.ErrorDesc(err))
	}

	errHeader := http.Header{"cache-control": []string{"no-cache"}}
	err2 := tmpl.Exec(req, resp, "error/error.html", status, errHeader, &struct {
		Meta       map[string]interface{} // placeholder, to make tmpl.Exec() happy
		StatusCode int
		Status     string
		Err        error
		tmpl.Common
	}{
		Meta:       make(map[string]interface{}),
		StatusCode: status,
		Status:     http.StatusText(status),
		Err:        err,
		Common:     tmpl.Common{ErrorID: errorID},
	})
	if err2 != nil {
		log15.Error("error during execution of error template", "error", err2, "error_id", errorID)
		err := fmt.Errorf("error during execution of error template (error id %v): %v", errorID, err2)
		if !handlerutil.DebugMode {
			err = errPublicFacingErrorMessage
		}
		http.Error(resp, err.Error(), http.StatusInternalServerError)
	}
}

// errPublicFacingErrorMessage is the public facing error message to display
// when not in debug mode (to hide potentially sensitive information in original
// error message). Normally, an HTML template is used to do this, but if rendering that fails,
// this is the plain-text backup.
var errPublicFacingErrorMessage = errors.New(`Sorry, thereâ€™s been a problem.

If this issue persists, please email us at support@sourcegraph.com.`)
