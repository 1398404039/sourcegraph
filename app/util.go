package app

import (
	"html/template"
	"net/http"

	"github.com/microcosm-cc/bluemonday"
)

func sanitizeHTML(html string) template.HTML {
	return template.HTML(bluemonday.UGCPolicy().Sanitize(html))
}

func copyRequest(r *http.Request) *http.Request {
	rCopy := *r
	urlCopy := *r.URL
	rCopy.URL = &urlCopy
	return &rCopy
}
