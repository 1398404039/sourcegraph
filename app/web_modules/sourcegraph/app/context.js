// @flow

const context: {
	appURL?: string;
	authorization?: string;
	cacheControl?: string;
	currentUser?: Object;
	userEmail?: string;
	currentSpanID?: string;
	userAgent?: string;
	assetsRoot?: string;
	buildVars?: Object;
	features?: {[key: string]: any};
} = {};

// Sets the values of the context given a JSContext object from the server.
export function reset(ctx: typeof context) {
	Object.assign(context, ctx);
}

export default context;
