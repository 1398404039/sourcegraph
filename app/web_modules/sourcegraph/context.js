let context;

// TODO(autotest) support window object.
if (typeof window !== "undefined") {
	context = {
		currentUser: window._currentUser,
		csrfToken: window._csrfToken,
		isMothership: window._isMothership,
		cacheControl: window._cacheControl || null,
	};
} else {
	context = {
		currentUser: null,
		csrfToken: "",
		isMothership: false,
		cacheControl: null,
	};
}

export default context;
