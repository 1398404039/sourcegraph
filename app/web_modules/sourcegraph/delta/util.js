// isDevNull returns if path is "/dev/null". That is the sentinel value in git diffs
// that indicates the file does not exist (i.e., it is a newly added file, or a removed file,
// if the old or new path is "/dev/null", respectively).
export function isDevNull(path) {
	return path === "/dev/null";
}
