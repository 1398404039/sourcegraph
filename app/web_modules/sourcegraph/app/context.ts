import {setGlobalFeatures} from "sourcegraph/app/features";
import {Features} from "sourcegraph/app/features";
import {setGlobalSiteConfig} from "sourcegraph/app/siteConfig";
import UserStore from "sourcegraph/user/UserStore";

class Context {
	csrfToken?: string;
	cacheControl?: string;
	currentSpanID?: string;
	userAgentIsBot?: boolean;

	// Some fields were migrated to React context from this global context object. These
	// getters prevent you from accidentally accessing these fields in their old home,
	// on this object.
	get currentUser(): void {
		throw new Error("currentUser is now accessible via this.context.user in components that specify 'user' in contextTypes");
	}
	get userEmail(): void {
		throw new Error("userEmail is no longer available globally; use the UserBackend/UserStore to retrieve it");
	}
	get hasLinkedGitHub(): void {
		throw new Error("hasLinkedGitHub is no longer available globally; use the UserBackend/UserStore directly");
	}
}

let context = new Context();

// ContextInput is the input context to set up the JS environment (e.g., from Go).
type ContextInput = typeof context & {
	// We are migrating from a global context object to using React context
	// as much as possible. These fields are only available using context wrappers.
	appURL?: string;
	assetsRoot?: string;
	buildVars?: {Version: string};
	features?: Features;

	// This is now available in UserStore.activeAccessToken.
	accessToken?: string;
};

// Sets the values of the context given a JSContext object from the server.
export function reset(ctx: ContextInput): void {
	const features = ctx.features;
	delete ctx.features;
	if (typeof features !== "undefined") {
		setGlobalFeatures(features);
	}

	const {appURL, assetsRoot, buildVars} = ctx;
	if (typeof appURL === "undefined" || typeof assetsRoot === "undefined" || typeof buildVars === "undefined") {
		throw new Error("appURL, assetsRoot, and buildVars must all be set");
	}
	setGlobalSiteConfig({appURL, assetsRoot, buildVars});
	delete ctx.appURL;
	delete ctx.assetsRoot;
	delete ctx.buildVars;

	if (ctx.accessToken) {
		UserStore.activeAccessToken = ctx.accessToken;
	}
	delete ctx.accessToken;

	Object.assign(context, ctx);
}

export default context;
