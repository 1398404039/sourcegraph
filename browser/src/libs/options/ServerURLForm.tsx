import { upperFirst } from 'lodash'
import * as React from 'react'
import { ERAUTHREQUIRED } from '../../../../shared/src/backend/errors'
import { EINVALIDSOURCEGRAPHURL } from '../../shared/util/context'
import { ConnectionStatus } from './useSourcegraphURL'

const statusClassNames: Record<ConnectionStatus['type'], string> = {
    connecting: 'bg-warning',
    connected: 'bg-success',
    error: 'bg-danger',
}

/**
 * This is the [Word-Joiner](https://en.wikipedia.org/wiki/Word_joiner) character.
 * We are using this as a &nbsp; that has no width to maintain line height when the
 * url is being updated (therefore no text is in the status indicator).
 */
const zeroWidthNbsp = '\u2060'

export interface ServerURLFormProps {
    className?: string
    connectionStatus: undefined | ConnectionStatus
    sourcegraphURL: string
    onSourcegraphURLChange: (value: string) => void
    onSourcegraphURLSubmit: () => void
    requestSourcegraphURLPermissions: (e: React.MouseEvent) => void
}

export const ServerURLForm: React.FunctionComponent<ServerURLFormProps> = ({
    connectionStatus,
    onSourcegraphURLChange,
    onSourcegraphURLSubmit,
    requestSourcegraphURLPermissions,
    sourcegraphURL,
    className,
}) => {
    const onURLChange = React.useCallback(
        ({ target: { value } }: React.ChangeEvent<HTMLInputElement>): void => {
            onSourcegraphURLChange(value)
        },
        [onSourcegraphURLChange]
    )
    const onURLSubmit = React.useCallback(
        (event: React.FormEvent<HTMLFormElement>): void => {
            event.preventDefault()
            onSourcegraphURLSubmit()
        },
        [onSourcegraphURLSubmit]
    )
    const statusString = connectionStatus ? upperFirst(connectionStatus.type) : zeroWidthNbsp
    const statusClassName = connectionStatus ? statusClassNames[connectionStatus.type] : 'bg-secondary'
    return (
        // eslint-disable-next-line react/forbid-elements
        <form className={'server-url-form ' + (className || '')} onSubmit={onURLSubmit}>
            <label htmlFor="sourcegraph-url">Sourcegraph URL</label>
            <div className="input-group">
                <div className="input-group-prepend">
                    <span className="input-group-text">
                        <span>
                            <span className={'server-url-form__status-indicator ' + statusClassName} />{' '}
                            <span className="e2e-connection-status">{statusString}</span>
                        </span>
                    </span>
                </div>
                <input
                    type="text"
                    className="form-control e2e-sourcegraph-url"
                    id="sourcegraph-url"
                    value={sourcegraphURL}
                    onChange={onURLChange}
                    spellCheck={false}
                    autoCapitalize="off"
                    autoCorrect="off"
                />
            </div>
            {connectionStatus &&
                connectionStatus.type === 'error' &&
                (connectionStatus.error.code === EINVALIDSOURCEGRAPHURL ? (
                    <div className="mt-1">Invalid Sourcegraph URL.</div>
                ) : connectionStatus.error.code === ERAUTHREQUIRED ? (
                    <div className="mt-1">
                        Authentication to Sourcegraph failed.{' '}
                        <a href={sourcegraphURL} target="_blank" rel="noopener noreferrer">
                            Sign in to your instance
                        </a>{' '}
                        to continue.
                    </div>
                ) : (
                    <div className="mt-1">
                        <p>
                            Unable to connect to{' '}
                            <a href={sourcegraphURL} target="_blank" rel="noopener noreferrer">
                                {sourcegraphURL}
                            </a>
                            . Ensure the URL is correct and you are{' '}
                            <a href={sourcegraphURL + '/sign_in'} target="_blank" rel="noopener noreferrer">
                                logged in
                            </a>
                            .
                        </p>
                        {connectionStatus.urlHasPermissions === false && (
                            <p>
                                You may need to{' '}
                                <a href="#" onClick={requestSourcegraphURLPermissions}>
                                    grant the Sourcegraph browser extension additional permissions
                                </a>{' '}
                                for this URL.
                            </p>
                        )}
                        <p>
                            <b>Site admins:</b> ensure that{' '}
                            <a
                                href="https://docs.sourcegraph.com/admin/config/site_config"
                                target="_blank"
                                rel="noopener noreferrer"
                            >
                                all users can create access tokens
                            </a>
                            .
                        </p>
                    </div>
                ))}
        </form>
    )
}
