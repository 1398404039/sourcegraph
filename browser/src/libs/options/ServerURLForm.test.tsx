import renderer from 'react-test-renderer'
import * as React from 'react'

import { ServerURLForm, ServerURLFormProps } from './ServerURLForm'
import { Omit } from 'utility-types'
import { InvalidSourcegraphURLError } from '../../shared/util/context'
import { AuthRequiredError } from '../../../../shared/src/backend/errors'

describe('ServerURLForm', () => {
    const props: Omit<ServerURLFormProps, 'sourcegraphURL' | 'connectionStatus'> = {
        onSourcegraphURLChange: jest.fn(),
        onSourcegraphURLSubmit: jest.fn(),
        requestSourcegraphURLPermissions: jest.fn(),
    }
    test('Editing', () => {
        expect(
            renderer
                .create(<ServerURLForm {...props} sourcegraphURL="https://sourcegraph" connectionStatus={undefined} />)
                .toJSON()
        ).toMatchSnapshot()
    })

    test('Connecting', () => {
        expect(
            renderer
                .create(
                    <ServerURLForm
                        {...props}
                        sourcegraphURL="https://sourcegraph.com"
                        connectionStatus={{ type: 'connecting' }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot()
    })

    test('Connected', () => {
        expect(
            renderer
                .create(
                    <ServerURLForm
                        {...props}
                        sourcegraphURL="https://sourcegraph.com"
                        connectionStatus={{ type: 'connected' }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot()
    })

    test('Invalid URL', () => {
        expect(
            renderer
                .create(
                    <ServerURLForm
                        {...props}
                        sourcegraphURL="sourcegraph.com"
                        connectionStatus={{ type: 'error', error: new InvalidSourcegraphURLError('sourcegraph.com') }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot()
    })

    test('Auth required', () => {
        expect(
            renderer
                .create(
                    <ServerURLForm
                        {...props}
                        sourcegraphURL="https://sourcegraph.sgdev.org"
                        connectionStatus={{
                            type: 'error',
                            error: new AuthRequiredError('https://sourcegraph.sgdev.org'),
                        }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot()
    })

    test('Other error', () => {
        expect(
            renderer
                .create(
                    <ServerURLForm
                        {...props}
                        sourcegraphURL="https://sourcegraph.sgdev.org"
                        connectionStatus={{
                            type: 'error',
                            error: new Error(),
                        }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot()
    })

    test('Other error - has permissions', () => {
        expect(
            renderer
                .create(
                    <ServerURLForm
                        {...props}
                        sourcegraphURL="https://sourcegraph.sgdev.org"
                        connectionStatus={{
                            type: 'error',
                            error: new Error(),
                            urlHasPermissions: true,
                        }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot()
    })

    test('Other error - no permissions', () => {
        expect(
            renderer
                .create(
                    <ServerURLForm
                        {...props}
                        sourcegraphURL="https://sourcegraph.sgdev.org"
                        connectionStatus={{
                            type: 'error',
                            error: new Error(),
                            urlHasPermissions: false,
                        }}
                    />
                )
                .toJSON()
        ).toMatchSnapshot()
    })
})
