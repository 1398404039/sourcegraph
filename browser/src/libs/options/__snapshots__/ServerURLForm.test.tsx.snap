// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServerURLForm Auth required 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-danger"
          />
           
          <span
            className="e2e-connection-status"
          >
            Error
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="https://sourcegraph.sgdev.org"
    />
  </div>
  <div
    className="mt-1"
  >
    Authentication to Sourcegraph failed.
     
    <a
      href="https://sourcegraph.sgdev.org"
      rel="noopener noreferrer"
      target="_blank"
    >
      Sign in to your instance
    </a>
     
    to continue.
  </div>
</form>
`;

exports[`ServerURLForm Connected 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-success"
          />
           
          <span
            className="e2e-connection-status"
          >
            Connected
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="https://sourcegraph.com"
    />
  </div>
</form>
`;

exports[`ServerURLForm Connecting 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-warning"
          />
           
          <span
            className="e2e-connection-status"
          >
            Connecting
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="https://sourcegraph.com"
    />
  </div>
</form>
`;

exports[`ServerURLForm Editing 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-secondary"
          />
           
          <span
            className="e2e-connection-status"
          >
            ‚Å†
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="https://sourcegraph"
    />
  </div>
</form>
`;

exports[`ServerURLForm Invalid URL 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-danger"
          />
           
          <span
            className="e2e-connection-status"
          >
            Error
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="sourcegraph.com"
    />
  </div>
  <div
    className="mt-1"
  >
    Invalid Sourcegraph URL.
  </div>
</form>
`;

exports[`ServerURLForm Other error - has permissions 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-danger"
          />
           
          <span
            className="e2e-connection-status"
          >
            Error
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="https://sourcegraph.sgdev.org"
    />
  </div>
  <div
    className="mt-1"
  >
    <p>
      Unable to connect to
       
      <a
        href="https://sourcegraph.sgdev.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://sourcegraph.sgdev.org
      </a>
      . Ensure the URL is correct and you are
       
      <a
        href="https://sourcegraph.sgdev.org/sign_in"
        rel="noopener noreferrer"
        target="_blank"
      >
        logged in
      </a>
      .
    </p>
    <p>
      <b>
        Site admins:
      </b>
       ensure that
       
      <a
        href="https://docs.sourcegraph.com/admin/config/site_config"
        rel="noopener noreferrer"
        target="_blank"
      >
        all users can create access tokens
      </a>
      .
    </p>
  </div>
</form>
`;

exports[`ServerURLForm Other error - no permissions 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-danger"
          />
           
          <span
            className="e2e-connection-status"
          >
            Error
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="https://sourcegraph.sgdev.org"
    />
  </div>
  <div
    className="mt-1"
  >
    <p>
      Unable to connect to
       
      <a
        href="https://sourcegraph.sgdev.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://sourcegraph.sgdev.org
      </a>
      . Ensure the URL is correct and you are
       
      <a
        href="https://sourcegraph.sgdev.org/sign_in"
        rel="noopener noreferrer"
        target="_blank"
      >
        logged in
      </a>
      .
    </p>
    <p>
      You may need to
       
      <a
        href="#"
        onClick={[MockFunction]}
      >
        grant the Sourcegraph browser extension additional permissions
      </a>
       
      for this URL.
    </p>
    <p>
      <b>
        Site admins:
      </b>
       ensure that
       
      <a
        href="https://docs.sourcegraph.com/admin/config/site_config"
        rel="noopener noreferrer"
        target="_blank"
      >
        all users can create access tokens
      </a>
      .
    </p>
  </div>
</form>
`;

exports[`ServerURLForm Other error 1`] = `
<form
  className="server-url-form "
  onSubmit={[Function]}
>
  <label
    htmlFor="sourcegraph-url"
  >
    Sourcegraph URL
  </label>
  <div
    className="input-group"
  >
    <div
      className="input-group-prepend"
    >
      <span
        className="input-group-text"
      >
        <span>
          <span
            className="server-url-form__status-indicator bg-danger"
          />
           
          <span
            className="e2e-connection-status"
          >
            Error
          </span>
        </span>
      </span>
    </div>
    <input
      autoCapitalize="off"
      autoCorrect="off"
      className="form-control e2e-sourcegraph-url"
      id="sourcegraph-url"
      onChange={[Function]}
      spellCheck={false}
      type="text"
      value="https://sourcegraph.sgdev.org"
    />
  </div>
  <div
    className="mt-1"
  >
    <p>
      Unable to connect to
       
      <a
        href="https://sourcegraph.sgdev.org"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://sourcegraph.sgdev.org
      </a>
      . Ensure the URL is correct and you are
       
      <a
        href="https://sourcegraph.sgdev.org/sign_in"
        rel="noopener noreferrer"
        target="_blank"
      >
        logged in
      </a>
      .
    </p>
    <p>
      <b>
        Site admins:
      </b>
       ensure that
       
      <a
        href="https://docs.sourcegraph.com/admin/config/site_config"
        rel="noopener noreferrer"
        target="_blank"
      >
        all users can create access tokens
      </a>
      .
    </p>
  </div>
</form>
`;
